### PR TITLE
8321939: [GenShen] ShenandoahOldEvacRatioPercent=100 fails with divide-by-zero

### PR DESCRIPTION
**Pull request by ysramakrishna**
The code that used the option ShenandoahOldEvacRatioPercent (SOEP) didn't correctly deal with the extremal setting of 100. Corrected the error and did some light refactoring in the vicinity, getting rid of some dead variables, adding documentation, and assertions. Also modified slightly the documentation of the option.

**Testing:**
- [x] GHA
- [x] code pipeline testing (TestPeriodicGC fails without the unrelated fix in https://github.com/openjdk/shenandoah/pull/372)
- [x] specjbb with SOEP=default,0,100 to verify that the issue is fixed

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8321939](https://bugs.openjdk.org/browse/JDK-8321939): [GenShen] ShenandoahOldEvacRatioPercent=100 fails with divide-by-zero (**Bug** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [a290be10](https://git.openjdk.org/shenandoah/pull/369/files/a290be10c58442aaac84bf6a08e0a4846bbe0fed)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/369/head:pull/369` \
`$ git checkout pull/369`

Update a local copy of the PR: \
`$ git checkout pull/369` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/369/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 369`

View PR using the GUI difftool: \
`$ git pr show -t 369`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/369.diff">https://git.openjdk.org/shenandoah/pull/369.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/369#issuecomment-1853127429)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -1181,91 +1181,100 @@ void ShenandoahHeap::cancel_old_gc() {
...
+  //     OE = old evac,
+  //     YE = young evac, and
+  //     TE = total evac = OE + YE
+  // By definition:
+  //            SOEP/100 = OE/TE
+  //                     = OE/(OE+YE)
+  //  => SOEP/(100-SOEP) = OE/((OE+YE)-OE)
+  //                     = OE/YE
+  //  =>              OE = YE*SOEP/(100-SOEP)
+
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1206]: Same comment as above, regarding the algebra steps.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 2]
```diff
@@ -1181,91 +1181,100 @@ void ShenandoahHeap::cancel_old_gc() {
...
-    reserve_for_mixed = 0;
   }
 
-  size_t available_for_promotions = max_old_reserve - reserve_for_mixed;
+  // Decide how much space we should reserve for promotions from young
+  size_t reserve_for_promo = 0;
+  const size_t promo_load = get_promotion_potential();
+  const bool doing_promotions = promo_load > 0;
   if (doing_promotions) {
     // We're only promoting and we have a maximum bound on the amount to be promoted
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1242]: With reference to the preceding comment, we should probably add an assert here: assert(!doing_mixed, "Don't do mixed and promotions in the same cycle")

Or maybe this is no longer true.  It doesn't have to be true.  The code that follows seems to suggest this is no longer true.  It was a simplification at some point in the implementation history.  If it is no longer true, let's remove that comment.

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp [hunk 3]
```diff
@@ -253,45 +252,31 @@ void ShenandoahGeneration::compute_evacuation_budgets(ShenandoahHeap* heap, bool
...
+  //
+  // Here's the algebra.
+  // Let SOEP = ShenandoahOldEvacRatioPercent,
+  //     OE = old evac,
+  //     YE = young evac, and
+  //     TE = total evac = OE + YE
+  // By definition:
+  //            SOEP/100 = OE/TE
+  //                     = OE/(OE+YE)
+  //  => SOEP/(100-SOEP) = OE/((OE+YE)-OE)
```
[the remark this answers is not in the document]
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:269]: ah, yes, I used the so-called "componendo-dividendo" rule from middle school, in this instantiation of the more general schema:

If a/b = c/d, then a/(b-a) = c/(d-c)

It is perhaps not very well-known in modern ratio-proportion pedagogy in grade-school, but is very useful for quick arithemtic.

https://hsm.stackexchange.com/questions/7544/etymology-of-some-terms-used-in-ratio-and-proportion-in-old-algebra-textbooks

[the remark this answers is not in the document]
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:269]: I'll add the justification in the comment.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 2 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1242]: I believe you are right. The comment appears obsolete. I'll check/confirm and delete the comment. Thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:269]: Thanks.  Your recollection of middle-school math is better than mine...  :)

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp:1]: Please ignore the changes in this file for now. It'll go away once the fix for the linked ticket https://bugs.openjdk.org/browse/JDK-8322150 is committed. This PR is blocked until that ticet's PR is committed.

**Comment by ysramakrishna** [comment]
This PR is blocked until PR https://github.com/openjdk/shenandoah/pull/372 integrates.

**Comment by ysramakrishna** [comment]
Converting to a draft until all failures are resolved satisfactorily.

**Comment by ysramakrishna** [comment]
This PR can be landed once the PR associated with JDK-8321816 integrates; the failures in GHA here have all been investigated and are either fixed by 8321816 or the result of changes from tip.

This PR is ready to commit in its current state.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 2 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp]: Done

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 1 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1207]: Done.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp]: Done.

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp:1]: This was backed out in a subsequent commit, so is not present in the final commit, as promised in previous comment.

**Pull request closed without merging**